include/mc_rtc/gui/Polyhedron.h

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2015-2019 CNRS-UM LIRMM, CNRS-AIST JRL
+ */
+
+#pragma once
+
+#include <mc_rtc/gui/details/traits.h>
+#include <mc_rtc/gui/elements.h>
+#include <mc_rtc/gui/types.h>
+
+namespace mc_rtc
+{
+
+namespace gui
+{
+
+/** Polyhedron should display a polyhedron or a set of polyhedrons
+ *
+ * A color is provided to display the polyhedron(s)
+ *
+ * \tparam GetT Should return an std::vector<Eigen::Vector3d> (one polyhedron) or an
+ * std::vector<std::vector<Eigen:Vector3d>> (list of polyhedrons)
+ *
+ */
+template<typename GetVerticesT>
+struct PolyhedronImpl : public Element
+{
+  static constexpr auto type = Elements::Polyhedron;
+
+  PolyhedronImpl(const std::string & name, const PolyhedronConfig & config, GetVerticesT get_vertices_fn)
+  : Element(name), config_(config), get_vertices_fn_(get_vertices_fn)
+  {
+    static_assert(details::CheckReturnType<GetVerticesT, PolyhedronTriangles>::value,
+                  "Polyhedron vertices callback must return an std::vector<Eigen::Vector3d> (triangle set with points "
+                  "ordered clockwise)");
```

Comment:
I think it is more common (and economic) to describe polyhedrons with two sets:
- a list of vertices (e.g. `std::vector<Eigen::Vector3d>`)
- a list of triangles (e.g. `std::vector<std::array<size_t, 3>>`)

I guess we can then use the config to discriminate whether the colors should be interpreted per triangle or per vertex

Reply:
Agreed it'd be more efficient. However most representations I've seen in the wild, and in particular the use-case of Julien and Hugo's polytopes describe them as `std::vector<std::vector<Eigen::Vector3d>>` (vector of triangles). I agree, this is not optimal, but for convenience sake I think it'd be nice to support such representations out of the box (especially considering that ROS provides a TRIANGLE_LIST element and that most of our visual rendering is currently based on ROS).

Maybe we can rename this element to `mc_rtc::gui::TriangleList` instead, and implement an `mc_rtc::gui::Polyhedron` with more optimal representations in the future?

Reply:
In [`e3b630b` (#296)](https://github.com/jrl-umi3218/mc_rtc/pull/296/commits/e3b630b01665d1cb82753f9696aa31d8d42e2271) I've added support for both formats with automatic conversion on the client side (so only one actual implementation has to be done)

The drawback for now is we have to specify `ColoredPolyhedron` to avoid ambiguities but I just had an idea to make it work easily without this thanks to a combination of `if constexpr` and auto return type.

If you are ok with this I'll merge once I have the work-around for the removal of `ColoredPolyhedron`

I also need to add an entry in dummy server that uses the other representation to make sure it works correctly in the client too.

Reply:
Wow thanks, that's perfect! :+1: 
LGTM when you're done :)